### PR TITLE
Extend support for getting the transaction script hash in std::tx

### LINKED ISSUE #2721: Extend support for getting the transaction script hash in std::tx
Since the operation involves knowing the constant values associated with the GTF opcode, it makes sense to move the following used function for the bridge predicate into the standard library itself:

```
use std::constants::ZERO_B256;

const GTF_SCRIPT_SCRIPT_LENGTH = 0x005;
const GTF_SCRIPT_SCRIPT = 0x00B;

/// Get the hash of the script bytecode
pub fn tx_script_bytecode_hash() -> b256 {
    // Get the script memory details
    let mut result_buffer: b256 = ZERO_B256;
    let script_length = __gtf::<u64>(0, GTF_SCRIPT_SCRIPT_LENGTH);
    let script_ptr = __gtf::<u64>(0, GTF_SCRIPT_SCRIPT);
    
    // Run the hash opcode for the script in memory
    asm(hash: result_buffer, ptr: script_ptr, len: script_length) {
        s256 hash ptr len;
        hash: b256
    }
}
```

### PR DESCRIPTION
**Pull request by pixelcircuits**
Adds new function for efficiently getting the hash of the script bytecode.
Resolves #2721

### CODE REVIEW

File: sway-lib-std/src/tx.sw [hunk 1]
```diff
@@ -5,6 +5,7 @@ use ::address::Address;
 use ::mem::read;
 use ::option::Option;
 use ::revert::revert;
+use ::constants::ZERO_B256;
```
Comment (mohammadfawaz) [inline sway-lib-std/src/tx.sw:8]: Do you mind keeping these imports in alphabetical order?

File: sway-lib-std/src/tx.sw [hunk 2]
```diff
@@ -232,6 +233,29 @@ pub fn tx_script_bytecode<T>() -> T {
...
 
+/// Get the hash of the script bytecode.
+/// Reverts if not a transaction-script
+pub fn tx_script_bytecode_hash() -> b256 {
+    let type = tx_type();
+    match type {
+        Transaction::Script => {
+            // Get the script memory details
+            let mut result_buffer: b256 = ZERO_B256;
+            let script_length = __gtf::<u64>(0, GTF_SCRIPT_SCRIPT_LENGTH);
```
Comment (mohammadfawaz) [inline sway-lib-std/src/tx.sw:244]: Can we use `tx_script_length()` here?

File: sway-lib-std/src/tx.sw [hunk 3]
```diff
@@ -232,6 +233,29 @@ pub fn tx_script_bytecode<T>() -> T {
...
+            let script_ptr = __gtf::<u64>(0, GTF_SCRIPT_SCRIPT);
```
Comment (mohammadfawaz) [inline sway-lib-std/src/tx.sw:245]: Can we use `tx_script_start_pointer()` here?

File: sway-lib-std/src/tx.sw [hunk 4]
```diff
@@ -232,6 +233,29 @@ pub fn tx_script_bytecode<T>() -> T {
...
+            
+            // Run the hash opcode for the script in memory
+            asm(hash: result_buffer, ptr: script_ptr, len: script_length) {
+                s256 hash ptr len;
+                hash: b256
+            }
+        },
+        _ => {
+            revert(0);
+        }
```
Comment (mohammadfawaz) [inline sway-lib-std/src/tx.sw:255]:
```suggestion
        _ => revert(0),
```

File: sway-lib-std/src/tx.sw [hunk 5]
```diff
@@ -232,6 +233,29 @@ pub fn tx_script_bytecode<T>() -> T {
...
+    match type {
+        Transaction::Script => {
+            // Get the script memory details
+            let mut result_buffer: b256 = ZERO_B256;
+            let script_length = __gtf::<u64>(0, GTF_SCRIPT_SCRIPT_LENGTH);
+            let script_ptr = __gtf::<u64>(0, GTF_SCRIPT_SCRIPT);
+            
+            // Run the hash opcode for the script in memory
+            asm(hash: result_buffer, ptr: script_ptr, len: script_length) {
+                s256 hash ptr len;
```
Comment (mohammadfawaz) [inline sway-lib-std/src/tx.sw:249]: I guess there is no way to use the `sha256` library here :/

File: test/src/sdk-harness/test_artifacts/tx_contract/src/main.sw [hunk 6]
```diff
@@ -2,22 +2,22 @@ contract;
 
 use std::{
     address::Address,
```
Comment (mohammadfawaz) [inline test/src/sdk-harness/test_artifacts/tx_contract/src/main.sw:4]: `Address` no longer needed here with the prelude :)

File: test/src/sdk-harness/test_projects/tx_fields/mod.rs [hunk 7]
```diff
@@ -170,6 +172,29 @@ async fn can_get_script_start_offset() {
     assert_eq!(result.value, script_start_offset as u64);
 }
 
+#[tokio::test]
+async fn can_get_script_bytecode_hash() {
```
Comment (Braqzen) [inline test/src/sdk-harness/test_projects/tx_fields/mod.rs:176]: Should there be another test which somehow checks the negative i.e. comparing one bytecode to another should produce unequal / different results? This proves that the call doesn't return the same output each time

File: sway-lib-std/src/tx.sw [hunk 8]
```diff
@@ -232,6 +233,29 @@ pub fn tx_script_bytecode<T>() -> T {
     read::<T>(tx_script_start_pointer())
 }
 
+/// Get the hash of the script bytecode.
+/// Reverts if not a transaction-script
+pub fn tx_script_bytecode_hash() -> b256 {
+    let type = tx_type();
+    match type {
```
Comment (Braqzen) [inline sway-lib-std/src/tx.sw:240]:
```suggestion
    match tx_type() {
```

Since the variable `type` is unused and doesn't cause indentation it is cleaner to include it in the `match` directly

File: sway-lib-std/src/tx.sw (hunk 5 above)
Comment (Braqzen) [reply sway-lib-std/src/tx.sw:249]: Why is that?

File: sway-lib-std/src/tx.sw [hunk 9]
```diff
@@ -232,6 +233,29 @@ pub fn tx_script_bytecode<T>() -> T {
...
+        Transaction::Script => {
+            // Get the script memory details
+            let mut result_buffer: b256 = ZERO_B256;
```
Comment (Braqzen) [inline sway-lib-std/src/tx.sw:243]:
```suggestion
            let mut result_buffer = ZERO_B256;
```

Compiler can infer on its own

File: sway-lib-std/src/tx.sw (hunk 8 above)
Comment (mohammadfawaz) [reply sway-lib-std/src/tx.sw:240]: I have a feeling this doesn't work but I could be wrong - I hope I'm wrong. If not, please file an issue @pixelcircuits

File: sway-lib-std/src/tx.sw (hunk 5 above)
Comment (mohammadfawaz) [reply sway-lib-std/src/tx.sw:249]: Mainly because all we have is a pointer and a length and not a concrete value that we can just pass to `sha256`. If we somehow can be build an array of bytes (`u8`s) out of this, then we can just use `sha256` directly, but I'm worried that's not possible just yet (and `Vec` doesn't help here because we'd only be hashing the struct itself and not the heap data).

Comment (pixelcircuits) [reply sway-lib-std/src/tx.sw:249]: @mohammadfawaz is correct. It's not currently possible to get an array of `u8`s through a simple compiler supported casting or anything. You could get an array of 64 bit "bytes", zero-padded to `u64` length, or the whole data rounded to the nearest `u64` with padded zeros at the end, but then this messes up the computed hash (and also is memory intensive). Since this is being used for bridging in a predicate that is likely to be used often, I think it makes sense to have an ultra optimized assembly defined function rather than try and fix current gaps.

File: sway-lib-std/src/tx.sw (hunk 8 above)
Comment (pixelcircuits) [reply sway-lib-std/src/tx.sw:240]: I don't think it works. Every function in the library does it this way (which I copy-pasted from). I'll test to verify and then file an issue if I can confirm it

File: test/src/sdk-harness/test_artifacts/tx_contract/src/main.sw (hunk 6 above)
Comment (pixelcircuits) [reply test/src/sdk-harness/test_artifacts/tx_contract/src/main.sw:4]: I get a compile error if I don't include this line, even after updating to the latest forc version

Comment (nfurfaro) [reply test/src/sdk-harness/test_artifacts/tx_contract/src/main.sw:4]: Not sure if the forc version with the prelude has been released yet... might be only on master?

File: test/src/sdk-harness/test_projects/tx_fields/mod.rs (hunk 7 above)
Comment (pixelcircuits) [reply test/src/sdk-harness/test_projects/tx_fields/mod.rs:176]: So it seems like all simple contract calls share the same script bytecode, so it's hard to test against different scripts. I do think it's a good idea that we check that the script is more than a byte long, which ensures that the script has some complexity to it with a unique hash we should expect

File: sway-lib-std/src/tx.sw (hunk 8 above)
Comment (pixelcircuits) [reply sway-lib-std/src/tx.sw:240]: So it works! Whatever issue there was before, it seems to have been fixed now

File: test/src/sdk-harness/test_artifacts/tx_contract/src/main.sw (hunk 6 above)
Comment (mohammadfawaz) [reply test/src/sdk-harness/test_artifacts/tx_contract/src/main.sw:4]: Wait, this test _is_ being added on `master`... why do we do need a release `forc`?

**Review by mohammadfawaz — APPROVED** [review]

File: sway-lib-std/src/tx.sw (hunk 8 above)
Comment (mohammadfawaz) [reply sway-lib-std/src/tx.sw]: Great! Glad to see these cleaned up.

**Comment by pixelcircuits** [comment]
@Braqzen Does this PR look good now?

**Review by Braqzen — APPROVED** [review]
I don't remember if I've said anything about `tx::*` as an import but I always complain about glob imports. Approving anyway
